gen/spec.go

```diff
@@ -432,3 +434,91 @@ func (p *printer) ok() bool { return p.err == nil }
 func tobaseConvert(b *BaseElem) string {
 	return b.ToBase() + "(" + b.Varname() + ")"
 }
+
+func (p *printer) varMapHeader(pre, post string, sizeVarname string, maxSize int) {
+
+	if maxSize <= 15 {
+		p.printf("\n%s 0x80 | (uint8(%s)&0x0F) %s", pre, sizeVarname, post)
+	} else if maxSize <= math.MaxUint16 {
```

Comment:
I think when `maxSize` is above 15, you should just fall back to emitting a call to `msgp.AppendMapHeader` (or its analog for encoding). The functional call overhead will be amortized over the large number of struct fields.

Reply:
Agreed - and the compiler might very well inline msgp.AppendMapHeader anyway.

Reply:
(I checked and the compiler does not appear to be inlining msgp.AppendMapHeader, at least in the case I looked at, but I agree the unnecessary bit operations here are more likely to result in an obscure bug than a measurable performance improvement).

Fixed.